### PR TITLE
Fix handling of system prompts with OpenAI's o1 models

### PR DESCRIPTION
**Pull request by dmontagu**
As reported on slack (https://pydanticlogfire.slack.com/archives/C083V7PMHHA/p1734734636385579), the current implementation results in issues with OpenAI's o1 models because they don't support `'role': 'system'`.

This PR ensures o1 models get system prompts as user prompts (as discussed here): https://community.openai.com/t/o1-models-do-not-support-system-role-in-chat-completion/953880/3). It also exposes a way to manually override the behavior, and specify to use role `'assistant'` rather than `'user'`, but I tentatively feel like `'user'` makes the most sense as the default (for o1 models; it remains as 'system' for others).

It also provides a way to disable system messages entirely if neither `'user'` or `'assistant'` feels right. I don't mind if we remove that option though if preferred.

I made additional changes in this PR to make the behavior easier to override; right now, to fix the behavior on main, you'd need to do:
```python
from __future__ import annotations as _annotations

from collections.abc import Iterable

from typing_extensions import assert_never

from pydantic_ai._utils import guard_tool_call_id as _guard_tool_call_id
from pydantic_ai.messages import (
    ModelRequest,
    RetryPromptPart,
    SystemPromptPart,
    ToolReturnPart,
    UserPromptPart,
)
from pydantic_ai.models import AgentModel, check_allow_model_requests
from pydantic_ai.models.openai import OpenAIModel, OpenAIAgentModel
from pydantic_ai.tools import ToolDefinition

try:
    from openai import NOT_GIVEN, AsyncOpenAI, AsyncStream
    from openai.types import ChatModel, chat
    from openai.types.chat import ChatCompletionChunk
    from openai.types.chat.chat_completion_chunk import ChoiceDeltaToolCall
except ImportError as _import_error:
    raise ImportError(
        'Please install `openai` to use the OpenAI model, '
        "you can use the `openai` optional group — `pip install 'pydantic-ai-slim[openai]'`"
    ) from _import_error


class OpenAIModel_O1(OpenAIModel):
    async def `agent_model`(
        self,
        *,
        function_tools: list[ToolDefinition],
        allow_text_result: bool,
        result_tools: list[ToolDefinition],
    ) -> AgentModel:
        check_allow_model_requests()
        tools = [self._map_tool_definition(r) for r in function_tools]
        if result_tools:
            tools += [self._map_tool_definition(r) for r in result_tools]
        return OpenAIAgentModel_O1(
            self.client,
            self.model_name,
            allow_text_result,
            tools,
        )


class OpenAIAgentModel_O1(OpenAIAgentModel):
    @classmethod
    def _map_user_message(cls, message: ModelRequest) -> Iterable[chat.ChatCompletionMessageParam]:
        for part in message.parts:
            if isinstance(part, SystemPromptPart):
                yield chat.ChatCompletionUserMessageParam(role='user', content=part.content)
            elif isinstance(part, UserPromptPart):
                yield chat.ChatCompletionUserMessageParam(role='user', content=part.content)
            elif isinstance(part, ToolReturnPart):
                yield chat.ChatCompletionToolMessageParam(
                    role='tool',
                    tool_call_id=_guard_tool_call_id(t=part, model_source='OpenAI'),
                    content=part.model_response_str(),
                )
            elif isinstance(part, RetryPromptPart):
                if part.tool_name is None:
                    yield chat.ChatCompletionUserMessageParam(role='user', content=part.model_response())
                else:
                    yield chat.ChatCompletionToolMessageParam(
                        role='tool',
                        tool_call_id=_guard_tool_call_id(t=part, model_source='OpenAI'),
                        content=part.model_response(),
                    )
            else:
                assert_never(part)
```

With the changes in this PR, if you wanted to change _back_ to passing a system message to an 'o1' model (since this PR addresses that by making it a user message), you'd still need to override the `OpenAIModel.agent_model` method in the same way, but the override to `OpenAIAgentModel._map_user_message` could be replaced with just:

```python
    def _map_model_request_part(self, part: ModelRequestPart) -> chat.ChatCompletionMessageParam:
        if self.model_name.startswith('o1') and isinstance(part, SystemPromptPart):
            return chat.ChatCompletionSystemMessageParam(role='system', content=part.content)
        return super()._map_model_request_part(part)
```

You wouldn't need to do that in practice because this PR adds the `system_role` parameter, but it will make it easier to override things in the future if something similar comes up for a different kind of message.

### CODE REVIEW

**Comment by dmontagu** [comment]
Alternatively, we could add a `system_prompt_role` parameter to `OpenAIModel` to make this easier to manually control. I'll open a PR doing that too and let @samuelcolvin decide what he prefers.

File: pydantic_ai_slim/pydantic_ai/models/openai.py [hunk 1]
```diff
@@ -268,30 +268,36 @@ def _map_message(cls, message: ModelMessage) -> Iterable[chat.ChatCompletionMess
...
-                yield chat.ChatCompletionUserMessageParam(role='user', content=part.content)
-            elif isinstance(part, ToolReturnPart):
-                yield chat.ChatCompletionToolMessageParam(
+            yield self._map_model_request_part(part)
+
+    def _map_model_request_part(self, part: ModelRequestPart) -> chat.ChatCompletionMessageParam:
+        if isinstance(part, SystemPromptPart):
+            if self.model_name.startswith('o1'):
+                # o1 does not currently support role 'system', so for now we use role 'user' instead
+                return chat.ChatCompletionUserMessageParam(role='user', content=part.content)
```
Comment (sydney-runkle) [inline pydantic_ai_slim/pydantic_ai/models/openai.py]: This makes sense 👍

**Comment by dmontagu** [comment]
I decided I think it makes more sense to add the ability to specify the `system_role`, as it will make it quicker to update things if/when openai adds support for system role to o1, or adds new models that also disallow the role `'system'`.

I'm okay going back to the way it was if we want to encourage people to do more subclassing/overriding of private stuff (or copy/pasting the whole model implementation..) to work around this, but I do think adding the `system_role` argument is a better option.

File: pydantic_ai_slim/pydantic_ai/models/openai.py [hunk 2]
```diff
@@ -90,8 +93,20 @@ def __init__(
                 [`AsyncOpenAI`](https://github.com/openai/openai-python?tab=readme-ov-file#async-usage)
                 client to use. If provided, `base_url`, `api_key`, and `http_client` must be `None`.
             http_client: An existing `httpx.AsyncClient` to use for making HTTP requests.
+            system_role: The role to use for SystemPromptPart messages. OpenAI's o1 models currently do not support
```
Comment (sydney-runkle) [inline pydantic_ai_slim/pydantic_ai/models/openai.py:96]:
```suggestion
            system_role: The role to use for `SystemPromptPart` messages. OpenAI's o1 models currently do not support
```

**Comment by sydney-runkle** [comment]
Seems reasonable!

**Comment by ArtemPolydom** [comment]
Guys you are wrong, they renamed "system" role to "developer"!

https://platform.openai.com/docs/guides/text-generation#developer-messages

**Comment by ArtemPolydom** [comment]
Make different PR and just rename "system" roles to "developer" for all openai models (developer role should work with old models too)

**Comment by dmontagu** [comment]
Closing and opening a new PR because I found some information about how this should be done properly https://platform.openai.com/docs/guides/text-generation#developer-messages looks like they updated to use "developer" messages

**Pull request closed without merging**